### PR TITLE
fix(portal): 修复交互式应用创建但是未启动完毕时无法连接问题

### PR DESCRIPTION
**Pull request by WZD09**

### CODE REVIEW

File: apps/portal-server/package.json [hunk 1]
```diff
@@ -38,7 +38,8 @@
     "ts-log": "2.2.5",
     "shell-quote": "1.8.0",
     "dayjs": "1.11.7",
-    "pino-pretty": "10.0.0"
+    "pino-pretty": "10.0.0",
+    "is-port-reachable": "^4.0.0"
```
Comment (ddadaal) [inline apps/portal-server/package.json:42]: 不要用^，用确定的版本`4.0.0`

**Review by ddadaal — APPROVED** [review]

**Comment by ddadaal** [comment]
is-port-reachable只分发es module格式的模块，但是我们的项目只支持commonjs格式的模块。看一下如何在commonjs里引用es module的模块：https://stackoverflow.com/questions/70396400/how-to-use-es6-modules-in-commonjs

File: apps/portal-server/src/clusterops/slurm/app.ts [hunk 2]
```diff
@@ -240,20 +240,44 @@ export const slurmAppOps = (cluster: string): AppOps => {
...
             // try to read the SESSION_INFO file to get port and password
               const infoFilePath = join(jobDir, SERVER_SESSION_INFO);
-              ready = await sftpExists(sftp, infoFilePath);
-
+              if (await sftpExists(sftp, infoFilePath)) {
+                const content = await sftpReadFile(sftp)(infoFilePath);
+                const serverSessionInfo = JSON.parse(content.toString()) as ServerSessionInfoData;
+                const { HOST, PORT } = serverSessionInfo;
+
+                import("is-port-reachable").then(async (isPortReachable) => {
```
Comment (ddadaal) [inline apps/portal-server/src/clusterops/slurm/app.ts:248]: import()返回的是一个Promise，可以用async await语法：

```ts
const isPortReachable = await import("is-port-reachable");
```

**Review by ddadaal — DISMISSED** [review]

**Comment by lyl-available** [comment]
这个需要提前配置什么吗？我直接切PR，界面报错了诶
image

{"level":50,"time":1682670765837,"pid":18,"hostname":"06c05aa27a27","req":"l","path":"/scow.portal.AppService/ListAppSessions","err":{"type":"Error","message":"require() of ES Module /app/node_modules/.pnpm/is-port-reachable@4.0.0/node_modules/is-port-reachable/index.js from /app/apps/portal-server/build/clusterops/slurm/app.js not supported.\nInstead change the require of index.js in /app/apps/portal-server/build/clusterops/slurm/app.js to a dynamic import() which is available in all CommonJS modules.","stack":"Error [ERR_REQUIRE_ESM]: require() of ES Module /app/node_modules/.pnpm/is-port-reachable@4.0.0/node_modules/is-port-reachable/index.js from /app/apps/portal-server/build/clusterops/slurm/app.js not supported.\nInstead change the require of index.js in /app/apps/portal-server/build/clusterops/slurm/app.js to a dynamic import() which is available in all CommonJS modules.\n    at /app/apps/portal-server/build/clusterops/slurm/app.js:205:105\n    at async /app/apps/portal-server/build/clusterops/slurm/app.js:205:57\n    at async Promise.all (index 4)\n    at async /app/apps/portal-server/build/clusterops/slurm/app.js:184:17\n    at async Object.listAppSessions (/app/apps/portal-server/build/clusterops/slurm/app.js:168:20)\n    at async Object.listAppSessions (/app/apps/portal-server/build/services/app.js:154:27)\n    at async augmentedImplementations.<computed> [as listAppSessions] (/app/node_modules/.pnpm/@ddadaal+tsgrpc-server@0.19.0_@grpc+grpc-js@1.8.14/node_modules/@ddadaal/tsgrpc-server/lib/server.js:76:33)","code":"ERR_REQUIRE_ESM"},"msg":"Error occurred. Return the error."}

**Comment by ddadaal** [comment]
应该还是在commonjs模块中导入esm模块的问题。感觉不好解决，看一下有没有其他类似的库。实在没有，就直接复制粘贴大法吧：这个is-port-reachable包实际上只有一个文件 （https://github.com/sindresorhus/is-port-reachable/blob/main/index.js ），而且实现很简单，可以把代码抄过来，放在src/utils/isPortReachable.ts下，并写一个注释指向源仓库。这个库是MIT协议的，复制粘贴没有问题

**Review by ddadaal — APPROVED** [review]

**Review by lyl-available — APPROVED** [review]

**Pull request merged**